### PR TITLE
Separate OpenMP from Pyccel's core functionalities and connect it as a Plugin

### PR DESCRIPTION
**Pull request by rmahjoubi**
This PR reintroduces OpenMP in Pyccel as a standalone plugin, interfacing with Pyccel via the plugins interface.

## OpenMP Plugin

**Classes**
- Treating TextX objects is done in the syntactic stage via classes that inherit from `OmpTxNode`, which itself inherits from `OmpNode`; visiting these classes results in instances of their respective classes that inherit from `OmpNode`
- Each OpenMP version has dedicated syntactic/semantic parsers and printers
- Printers use the raw statement of an OpenMP object provided in the comment; the segmentation of these statements allows for the suppression or tweaking of printing of segments correlated to the desired OpenMP object

## Commit Summary

- Allow the use of protected access in plugins as their contents are designed to be sub-class functions
- Update docs to make them version agnostic
- Remove mentions to OpenMP in the core code
- Add a new hook `remove_cli_arguments` to remove abandoned CLI kwargs.
- Group arguments of `epyccel_seq` into a `kwargs` object (the docstrings are still available in the user-facing function `epyccel`) but grouping allows plugin tweaks like adding new accelerators
- Add the OpenMP plugin implementing
    - Dedicated AST nodes
    - Dedicated grammar
    - Syntactic methods
    - Semantic methods
    - Printing methods
- Call `plugin.read_cli_arguments` earlier in `handle_plugin_arguments` to pop irrelevant arguments
- Make numpy global in tests
- Add new tests
    - `test_omp_target_teams_distribute_parallel_for`
    - `potential_internal_data_race_condition`
    - `parallel_if`
    - `stenc_2d`
- Update existing tests 
- Deprecate `#$ omp end for`
- Add tests to ensure errors are raised when parsing OpenMP pragmas.
- Remove `test_preprocess`. The results are not checked so calling `parse` successfully doesn't tell us anything.
- Add `# $ omp end parallel` to syntax check as matching start and end of parallel is now checked at syntax level.
- Pass the initialised OpenMP plugin to any tests calling `Parser` explicitly.

---

## Pull request checklist

Please tick off items when you have completed them or determined that they are not necessary for this pull request:

- [x] Write a clear PR description
- [x] Ensure you appear in the AUTHORS file
- [x] Add tests to check your code works as expected
- [x] Update documentation if necessary
- [x] Update CHANGELOG.md
- [x] Ensure any relevant issues are linked
- [x] Ensure new tests are passing

### CODE REVIEW

**Comment by EmilyBourne** [comment]
> > Deregister?
> 
> Didn't work, should i just add unregister and unpatch to dict_custom

I prefer "deregister" (https://english.stackexchange.com/a/40095) but I think you can add this to `dict_custom`

I have no idea what is meant by "unpatch" though. Maybe just say "Removing patches" ?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
I think this PR is nearly ready for review so I have gone ahead and done a preliminary review.

After a first look through this really looks great! Congratulations on implementing this so cleanly. There are a couple of points that aren't fully generic but the vast majority seems flexible and easily extendable in the future. Well done :smiley:

**Comment by rmahjoubi** [comment]
@EmilyBourne Thanks for the review! Regarding the current functionality of the `register` and `deregister` methods in the `PluginsManager`:

`register`: This method is intended to be called from within Pyccel’s internal classes that we want to expose to plugins. It invokes the `register` method implemented by each plugin, which should add the relevant instances to a registry and apply the necessary patches for the first time.

`deregister`: This method calls the `deregister` function implemented by each specified plugin. It removes the corresponding instances from the registry and undoes the applied patches.

The idea behind the registry was to introduce a caching behavior — for example, to avoid reapplying patches every time epyccel is called, which is often unnecessary.

However, I think we can put that aside for now and go with the simpler approach: removing the registry, `deregister`, and `refresh`  methods altogether.

**Comment by EmilyBourne** [comment]
> @EmilyBourne Thanks for the review! Regarding the current functionality of the `register` and `deregister` methods in the `PluginsManager`:
> 
> `register`: This method is intended to be called from within Pyccel’s internal classes that we want to expose to plugins. It invokes the `register` method implemented by each plugin, which should add the relevant instances to a registry and apply the necessary patches for the first time.
> 
> `deregister`: This method calls the `deregister` function implemented by each specified plugin. It removes the corresponding instances from the registry and undoes the applied patches.
> 
> The idea behind the registry was to introduce a caching behavior — for example, to avoid reapplying patches every time epyccel is called, which is often unnecessary.
> 
> However, I think we can put that aside for now and go with the simpler approach: removing the registry, `deregister`, and `refresh`  methods altogether.

Maybe I'm missing things/have misunderstood but wouldn't that cause problems with epyccel? If you call it once with a plugin and can't remove it then you won't have the expected behaviour on the second call.

I think it looks good as is, I would just prefer a more accurate name.

Can you point me at where the registry is managed. Maybe I need another look at this

File: developer_docs/plugins.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
+# Pyccel Plugin System
+
+This document provides an overview of the Pyccel plugin system and a guide on how to create new plugins.
+
+## Overview
+
+The Pyccel plugin system allows for extending Pyccel's functionality through plugins. Plugins can modify the behaviour of existing classes by patching their methods or adding new methods. This is particularly useful for adding support for new language features, optimisations, or code generation targets without modifying the core codebase.
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply developer_docs/plugins.md]: fixed 05f8336

File: developer_docs/plugins.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+The plugin system is built around the following key components:
+
+1. `Plugin` - Abstract base class that all plugins must implement
+2. `Plugins` - Singleton manager that handles loading, registering, and deregistering plugins
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply developer_docs/plugins.md]: fixed a5abcbd

File: developer_docs/plugins.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        
+    @abstractmethod
+    def register(self, instances):
+        """Register instances with this plugin."""
+        pass
+        
+    @abstractmethod
+    def refresh(self):
+        """Refresh all registered targets."""
+        pass
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply developer_docs/plugins.md:33]: This method is called to refresh all targets that have been registered with this plugin. It should reapply patches or apply new patches if necessary to ensure that the plugin's functionality is up to date

File: developer_docs/plugins.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        
+        # Call the original method
+        result = original_method(self, *args, **kwargs)
+        
+        # Do something after the original method
+        print("After parsing")
+        
+        return result
+    
+    # Apply the patch
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply developer_docs/plugins.md:168]: it's up to the plugin implementation, I defined it here for simplicity.

File: pyccel/parser/extend_tree.py [hunk 5]
```diff
@@ -45,7 +68,7 @@ def get_comments(code):
     if comments:
         new_comments    = [[comments[0]]]
         for index in range(1,len(comments)):
-            if comments[index][1] == comments[index-1][1]+1 and comments[index][2] == comments[index-1][2]:
+            if comments[index][1] == comments[index-1][1]+1 and comments[index][2] == comments[index-1][2] and not re.match(r'^#\$ *omp', comments[index - 1][0]):
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/parser/extend_tree.py:72]: I think a better alternative is not grouping comments that represent pragmas in a CommentMultiLine.
what do you think

File: pyccel/parser/syntactic.py [hunk 6]
```diff
@@ -233,16 +232,14 @@ def _treat_comment_line(self, line, stmt):
...
-                expr = omp_parse(stmts=line)
-                try:
-                    expr = omp_parse(stmts=line)
-                except TextXSyntaxError as e:
-                    errors.report(f"Invalid OpenMP header. {e.message}",
-                            symbol = stmt, column = e.col,
-                              severity='fatal')
+                errors.report("OpenMP support is disabled. To enable it, please re-run the program with the --openmp flag (or epyccel with accelerators=['openmp']).",
+                              symbol = stmt,
+                              severity='warning')
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/parser/syntactic.py]: fixed 127237a

File: pyccel/plugins/Openmp/omp.py [hunk 7]
```diff
@@ -0,0 +1,1139 @@
...
+    >>> node.lineno
+    10
+    >>> node.col_offset
+    5
+    """
+
+    def __init__(self, lineno, col_offset):
+        super().__init__()
+        self.lineno = lineno
+        self.col_offset = col_offset
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/plugins/Openmp/omp.py:65]: I just needed a class that inherits from AST where i can store the line and column of the openmp statement when it's first created after parsing with textx, should i change the doc string

File: pyccel/plugins/Openmp/omp.py [hunk 8]
```diff
@@ -0,0 +1,1139 @@
...
+    It provides common functionality for handling OpenMP directives and constructs.
+
+    Parameters
+    ----------
+    raw : str
+        The raw syntax of the comment's section that represents the object, 
+        as provided in the source code.
+    position : tuple
+        The start and end positions of the OpenMP syntax in the source code, 
+        used to print errors.
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/plugins/Openmp/omp.py:82]: it's necessary in this case since a pragma like :  `#$ omp for reduction(+:func_result)` is segmented into multiple nodes (directive, clause ..) each have a start and an end position.

File: pyccel/plugins/Openmp/omp.py [hunk 9]
```diff
@@ -0,0 +1,1139 @@
...
+    OmpDirective : Class representing the directive that starts a construct.
+    OmpEndDirective : Class representing the directive that ends a construct.
+
+    Examples
+    --------
+    >>> from pyccel.ast.core import CodeBlock
+    >>> start_directive = OmpDirective(name="parallel", clauses=(), is_construct=True,
+    ...                             raw="omp parallel", position=(4, 17))
+    >>> end_directive = OmpEndDirective(name="parallel", clauses=(), is_construct=True,
+    ...                             raw="omp end parallel", position=(4, 21))
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/plugins/Openmp/omp.py:175]: yes the parallel construct needs an end directive

File: pyccel/plugins/Openmp/openmp_4_5.py [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+                directive = OmpTxEndDirective(model.statement, line, cls._version, lineno=expr.lineno,
+                                              column=expr.col_offset) if model.statement.is_end_directive else OmpTxDirective(
+                    model.statement, line, cls._version, lineno=expr.lineno, column=expr.col_offset)
+                return instance._visit(directive)
+            except TextXError as e:
+                errors.report(e.message, severity="fatal", symbol=expr)
+        return method(line, expr)
+
+    @staticmethod
+    def _visit(instance, stmt, method, cls=None):
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/plugins/Openmp/openmp_4_5.py:250]: the actual method being called is this one:
https://github.com/pyccel/pyccel/blob/2943ba498e00d08ee71027bfbf3cff849c315fb6/pyccel/plugins/Openmp/openmp_4_5.py#L91-L100

File: pyccel/plugins/Openmp/plugin.py [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+        False
+        """
+        for method_name in list(self.patches.keys()):
+            self.patches[method_name] = [
+                patch for patch in self.patches[method_name]
+                if patch.version != version
+            ]
+            if not self.patches[method_name]:
+                del self.patches[method_name]
+        self.loaded_versions.remove(version)
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply pyccel/plugins/Openmp/plugin.py:173]: I am not sure which class variable ?

File: tests/epyccel/modules/openmp.py [hunk 12]
```diff
@@ -207,10 +209,17 @@ def test_omp_get_initial_device():
...
+    #$ omp target teams distribute parallel for reduction(+:x) map(always,tofrom:x)
+    for i in range(10):
+        x += i
+    return x
+
 def test_omp_get_set_schedule():
     import numpy as np
     from pyccel.stdlib.internal.openmp import omp_get_schedule, omp_set_schedule
     func_result = np.int32(0)
+    i = 0 # pylint: disable=unused-variable
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply tests/epyccel/modules/openmp.py]: fixed 0e3acdb

File: tests/errors/syntax_blockers/ex1.py
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply tests/errors/syntax_blockers/ex1.py:1]: it tests that a syntax error is raised if a comment starts with `#$` . now we only raise a warning

**Comment by rmahjoubi** [comment]
/bot run pyccel_lint docs spelling pylint intel

**Comment by rmahjoubi** [comment]
/bot run pyccel_lint docs spelling pylint intel

**Comment by EmilyBourne** [comment]
I can reproduce but the error doesn't occur every time I run. There is a warning coming from this test. I wonder if it is related:
```
  /home/emily/Code/pyccel/pyccel/codegen/compiling/compilers.py:589: UserWarning: ifx: command line warning #10157: ignoring option '-W'; argument is of wrong type
  /home/emily/Code/pyccel/tests/epyccel/__epyccel__/__pyccel__/mod_pti80lcg9avg.f90(34): warning #9151: The DEPEND clause on the OpenMP* ORDERED directive is deprecated in favor of the DOACROSS clause.
          !$omp ordered depend(sink: i, j-1) depend(sink: i-1, j)
  -----------------------------^
  /home/emily/Code/pyccel/tests/epyccel/__epyccel__/__pyccel__/mod_pti80lcg9avg.f90(34): warning #9151: The DEPEND clause on the OpenMP* ORDERED directive is deprecated in favor of the DOACROSS clause.
          !$omp ordered depend(sink: i, j-1) depend(sink: i-1, j)
  --------------------------------------------------^
  /home/emily/Code/pyccel/tests/epyccel/__epyccel__/__pyccel__/mod_pti80lcg9avg.f90(37): warning #9151: The DEPEND clause on the OpenMP* ORDERED directive is deprecated in favor of the DOACROSS clause.
          !$omp ordered depend(source)
  -----------------------------^
```

**Comment by EmilyBourne** [comment]
Using `doacross` doesn't fix the error

**Comment by EmilyBourne** [comment]
The code is almost identical to the example here:
https://passlab.github.io/Examples/contents/Chap_synchronization/10_Doacross_Loop_Nest.html

So I wonder if it is a compiler bug

**Comment by EmilyBourne** [comment]
I tested with:
ifx (IFX) 2025.2.0 20250605
Copyright (C) 1985-2025 Intel Corporation. All rights reserved.

**Comment by EmilyBourne** [comment]
@rmahjoubi The `collapse(2)` seems to be the source of the problems. You have `ordered(2)`

> An ordered clause can be used on a loop construct with an integer parameter argument to define the number of associated loops within a doacross loop nest where cross-iteration dependences exist.

Collapse collapses your 2 loops into 1 loop so it is probably tricky for ifx to handle something with 2 associated loops when it only has 1 loop as input. Removing the collapse seems to fix the issue. I don't know if this is a compiler bug or a specification corner case.

File: tests/epyccel/modules/openmp.py [hunk 13]
```diff
@@ -489,3 +508,47 @@ def omp_long_line(long_variable_1_oiwed423rnoij21d4kojklm : 'int[:]', long_varia
...
+    for i in range(l_start, l_end):
+        a[i] = 2 * i
+    #$ omp end parallel
+    return a
+
+def stenc_2d(matrix: 'int[:,:]', n: 'int', m: 'int'):
+    from pyccel.stdlib.internal.openmp import omp_set_num_threads
+    omp_set_num_threads(np.int32(4))
+    #$ omp parallel
+    #$ omp for ordered(2) collapse(2)
```
[the remark this answers is not in the document]
Comment (rmahjoubi) [reply tests/epyccel/modules/openmp.py]: I tried going over the openmp 4.5 specification to see if there are any restrictions but couldn't find any, the only one i found is:

>If both the collapse and ordered clause with a parameter are specified, the parameter of the
>ordered clause must be greater than or equal to the parameter of the collapse clause.


so  it could be a compiler bug

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/modules/openmp.py]: Sounds like it in that case. We should open a bug report

**Comment by EmilyBourne** [comment]
@rmahjoubi 
image

File: pyccel/plugins/Openmp/grammar/openmp.tx [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+
+////////////////////////////////////////////////////
+//              Clauses values
+////////////////////////////////////////////////////
+atomic_clause_values: 'read' | 'write' | 'update' | 'capture';
+OmpTxScalarExpr: value=InnerContent | ParenthesisContent;
+OmpTxIntegerExpr: value=InnerContent | ParenthesisContent;
+OmpTxConstantPositiveInteger: value=/[0-9]+/;
+OmpTxList: throw_away+=ID[','];
+ParenthesizedExpression: (('(' ParenthesizedExpression ')') | /[^()]*/)*; // TODO: find a better way to do this as this will not escape the parenthesis inside of a string
```
Comment (EmilyBourne) [inline pyccel/plugins/Openmp/grammar/openmp.tx:408]: Can you elaborate more on this TODO? Maybe I can help? Usually `.tx` files are designed to handle parsing in a way that is more explicit than regexes to avoid this kind of issue.

Comment (EmilyBourne) [reply pyccel/plugins/Openmp/grammar/openmp.tx:408]: Maybe you are just missing something that matches a string?

File: pyccel/plugins/Openmp/grammar/openmp.tx [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+DataSharingAttribute: ('shared' | 'none' | 'private' | 'firstprivate');
+ReductionOperator: ('+' | '-' | '*' | '&' | '|' | '^' | '&&' | '||' | 'min' | 'max'); // TODO: those are only the implicitly declared ones for C and C++. Add the others for Fortran and support user-defined ones (see specs https://www.openmp.org/spec-html/5.0/openmpsu107.html#x140-581002)
```
Comment (EmilyBourne) [inline pyccel/plugins/Openmp/grammar/openmp.tx:410]: Forgotten TODO

File: pyccel/plugins/Openmp/omp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/omp.py:65]: Ah! Because comments don't exist in `ast.AST`? The docstring is fine I just didn't understand at first why a subclass was needed.

File: pyccel/plugins/Openmp/omp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/omp.py:82]: Ok. Please add a comment about this

File: pyccel/plugins/Openmp/plugin.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/plugin.py:173]: Me neither :sweat_smile:

**Review by EmilyBourne — DISMISSED** [review]
Do you think there is anyway to split this PR? It really is enormous and quite hard to review. I would love to split it into a Plugin part and an OpenMP part but I guess that would be tricky for testing? Could we have fake plugin for the tests in `tests/plugin` to get round this?

File: pyccel/plugins/Openmp/openmp_4_5/semantic.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+    return self._visit_for_construct(expr)
+
+
+def _visit_target_teams_distribute_parallel_for_construct(self, expr):
+    return self._visit_for_construct(expr)
+
+
+def _visit_OmpEndDirective(self, expr):
+    if not isinstance(expr.current_user_node, OmpConstruct) and expr.is_construct:
+        if expr.name == "for":
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/openmp_4_5/semantic.py:89]: /bot accept
Deprecation warning tested locally

File: pyccel/plugins/Openmp/openmp_4_5/semantic.py [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+            errors.report(
+                "Support for `#$omp end for` is deprecated and will be removed in v2.5. The end directive is auto-inserted at the end of the for loop",
+                symbol=expr,
+                severity="warning",
+            )
+            return EmptyNode()
+        else:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/openmp_4_5/semantic.py:96]: /bot accept
Deprecation warning tested locally

File: pyccel/plugins/Openmp/openmp_4_5/syntactic.py [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+        if txt.endswith("&") or txt.endswith("\\"):
+            if txt.endswith("&"):
+                errors.report(
+                    "Using & as an OpenMP continuation character is deprecated and will be removed in v2.5.",
+                    severity="warning",
+                    symbol=expr,
+                )
+            if self._multiline_directive_in_progress:
+                self._multiline_directive_in_progress.append(txt[3:-1])
+            else:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/openmp_4_5/syntactic.py:166]: /bot accept
Validated by #2633 (requires a 3 line pragma to test)

File: pyccel/plugins/Openmp/plugin.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+
+    @property
+    def version(self):
+        """
+        Get the OpenMP version.
+
+        Get the OpenMP version.
+        """
+        return self._version
+
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/plugin.py:27]: /bot accept
Utility function

File: pyccel/plugins/Openmp/openmp_4_5/ccode.py [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+def _print_OmpDirective(self, expr):
+    return f"#pragma omp {expr.raw}\n"
+
+
+def _print_OmpEndDirective(self, expr):
+    if isinstance(expr.current_user_node, OmpConstruct):
+        return ""
+    else:
+        return f"#pragma omp end {expr.raw}\n"
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/openmp_4_5/ccode.py:35]: /bot accept
I assume this else is included for completeness.
Unless @rmahjoubi can shed light on what it was designed to handle

File: pyccel/plugins/Openmp/plugin.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+        Functions to be added or to override codegen methods.
+    """
+    if language == "fortran":
+        return [getattr(openmp_config.fcode, n) for n in openmp_config.fcode.__all__]
+    elif language == "c":
+        return [getattr(openmp_config.ccode, n) for n in openmp_config.ccode.__all__]
+    elif language == "python":
+        return [getattr(openmp_config.pycode, n) for n in openmp_config.pycode.__all__]
+    else:
+        return []
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/plugins/Openmp/plugin.py:180]: /bot accept
No methods are added for unknown languages. There are no tests for unsupported languages but they could appear in other plugins

File: pyccel/plugins/Openmp/ast/omp.py [hunk 22]
```diff
@@ -0,0 +1,1217 @@
...
+        }
+        omp_exprs = getattr(tx_clause, "omp_exprs", tuple())
+        if not isinstance(omp_exprs, tuple):
+            omp_exprs = (omp_exprs,) if omp_exprs else tuple()
+        omp_exprs = tuple(
+            globals().get(ex.__class__.__name__)(
+                ex, comment, omp_version, lineno=lineno, column=column
+            )
+            for ex in omp_exprs
+        )
```
Comment (EmilyBourne) [inline pyccel/plugins/Openmp/ast/omp.py:983]: I don't understand why this is saved to a tuple. It is either a `OmpTxList` or a `OmpTxScalarExpr`. If it is a list would it not make more sense to unpack it into the tuple rather than having a tuple with 1 element that is a list?

File: pyccel/plugins/Openmp/openmp_4_5/syntactic.py [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+    return self._visit_for_directive(expr)
+
+
+def _visit_target_teams_distribute_parallel_for_directive(self, expr):
+    return self._visit_for_directive(expr)
+
+
+def _visit_OmpTxClause(self, expr):
+    omp_exprs = tuple(self._visit(e) for e in expr.omp_exprs)
+    return OmpClause(omp_exprs=omp_exprs, **expr.get_fixed_state())
```
Comment (EmilyBourne) [inline pyccel/plugins/Openmp/openmp_4_5/syntactic.py:311]: `OmpTxClause` contains lots more information than `OmpClause`. It is a shame to loose this information (e.g. for a reduction with `OmpTxClause` I can do `expr.operator` but this information is only found in the raw string in `OmpClause` this makes it hard to fix #2193. Either a sub-class `OmpReduction` should be added in a later PR to store extra information and/or an `OmpClause` should retain a reference to the `OmpTxClause`)

File: pyccel/plugins/Openmp/ast/omp.py
Comment (EmilyBourne) [inline pyccel/plugins/Openmp/ast/omp.py:1]: There are no slots in this file

File: pyccel/plugins/Openmp/ast/omp.py [hunk 24]
```diff
@@ -0,0 +1,1217 @@
...
+
+    See Also
+    --------
+    OmpTxDirective : Parent class for TextX-based OpenMP directives.
+    OmpEndDirective : Parent class for OpenMP end directives.
+    OmpConstruct : Class that uses end directives to delimit blocks of code.
+    """
+
+
+class OmpTxClause(OmpTxNode, OmpClause):
```
Comment (EmilyBourne) [inline pyccel/plugins/Openmp/ast/omp.py:937]: Is the double inheritance helpful?

**Comment by EmilyBourne** [comment]
I'm putting this back in draft as there seems to be a lot of information lost between the textx parsing stage and the Pyccel AST storage
